zookeeper/component.py

```diff
@@ -0,0 +1,364 @@
+from inspect import getmro, isclass
+
+from prompt_toolkit import print_formatted_text
+from typeguard import check_type
+from zookeeper.utils import (
+    convert_to_snake_case,
+    get_concrete_subclasses,
+    prompt_for_component,
+    promt_for_param_value,
+)
+
+try:  # pragma: no cover
+    from colorama import Fore
+
+    BLUE, YELLOW, RESET = Fore.BLUE, Fore.YELLOW, Fore.RESET
+except ImportError:  # pragma: no cover
+    BLUE = YELLOW = RESET = ""
+
+# Indent for nesting in the string representation
+INDENT = " " * 4
+
+
+def defined_on_self_or_parent(self, name):
+    # Using `hasattr` is not safe, as it is implemented with `getattr` wrapped
+    # in a try-catch (Python is terrible), so we need to check `dir(self)`.
+    if name in dir(self):
+        return True
+    if self.__component_parent__:
+        return defined_on_self_or_parent(self.__component_parent__, name)
+    return False
+
+
+def str_key_val(key, value, color=True, single_line=False):
+    if isinstance(value, Component):
+        if single_line:
+            value = repr(value)
+        else:
+            value = f"\n{INDENT}".join(str(value).split("\n"))
+    elif callable(value):
+        value = "<callable>"
+    elif type(value) == str:
+        value = f'"{value}"'
+    space = "" if single_line else " "
+    return (
+        f"{BLUE}{key}{RESET}{space}={space}{YELLOW}{value}{RESET}"
+        if color
+        else f"{key}{space}={space}{value}"
+    )
+
+
+class Component:
+    """
+    A generic, modular component class designed to be easily configurable.
+
+    Components can have configurable parameters, which can be either generic
+    Python objects or nested sub-components. These are declared with class-level
+    Python type annotations, in the same way that elements of
+    [dataclasses](https://docs.python.org/3/library/dataclasses.html) are
+    declared. After instantiation, components are 'configured' with a
+    configuration dictionary; this process automatically injects the correct
+    parameters into the component and all subcomponents. Component parameters
+    can have defaults set, either in the class definition or passed via
+    `__init__`, but configuration values passed to `configure` will always take
+    precedence over these values.
+
+    If a nested sub-component child declares a parameter with the same name as a
+    parameter in one of its ancestors, it will receive the same configured value
+    as the parent does. Howevever, configuration is scoped: if the parameter on
+    the child, or on a _closer anscestor_, is configured with a different value,
+    then that value will override the one from the original parent.
+
+    Hydration can be interactive. In this case, the method will prompt for
```

Comment:
If I recall correctly, you changed the name in the code?
```suggestion
    Configuration can be interactive. In this case, the method will prompt for
```